File: clients/php/tests/SchedulerTest.php
<?php
declare(strict_types=1);
namespace Vexilla\Tests;

use Carbon\Carbon;
use PHPUnit\Framework\TestCase;
use Vexilla\Scheduler;
use Vexilla\Types\Schedule;
use Vexilla\Types\ScheduleTimeType;
use Vexilla\Types\ScheduleType;

final class SchedulerTest extends TestCase
{

  public function testActiveNone()
  {
    $schedule = new Schedule(0, 0, 'UTC', ScheduleTimeType::NONE, 0, 0);
    $this->assertTrue(Scheduler::isScheduleActive($schedule, ScheduleType::EMPTY ));
  }
  public function testActiveStartEndSingleDay()
  {
    $now = Carbon::now();
    $zeroDay = Carbon::createFromTimestampMsUTC(0);

    for ($hour = 0; $hour < 24; $hour++) {

      $mockedNow = Carbon::create(
        $now->year,
        $now->month,
        $now->day,
        $hour,
        0,
        0,
      );

      $zeroDayWithMockedTime = $zeroDay->clone();
      $zeroDayWithMockedTime->setHours($hour);

      $beforeSchedule = new Schedule(
        $mockedNow->timestamp * 1000,
        $mockedNow->timestamp * 1000,
        'UTC',
        ScheduleTimeType::START_END,
        $zeroDayWithMockedTime->clone()->addHours(1)->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHours(3)->timestamp * 1000,
      );

      $isBeforeScheduleActive = Scheduler::isScheduleActiveWithNow($beforeSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isBeforeScheduleActive, "Hour $hour: isBeforeScheduleActive");

      $duringSchedule = new Schedule(
        $mockedNow->timestamp * 1000,
        $mockedNow->timestamp * 1000,
        'UTC',
        ScheduleTimeType::START_END,
        $zeroDayWithMockedTime->clone()->subHour()->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHour()->timestamp * 1000,
      );

      $isDuringScheduleActive = Scheduler::isScheduleActiveWithNow($duringSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertTrue($isDuringScheduleActive, "Hour $hour: isDuringScheduleActive");


      $afterSchedule = new Schedule(
        $mockedNow->timestamp * 1000,
        $mockedNow->timestamp * 1000,
        'UTC',
        ScheduleTimeType::START_END,
        $zeroDayWithMockedTime->clone()->subHours(3)->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->subHours(1)->timestamp * 1000,
      );

      $isAfterScheduleActive = Scheduler::isScheduleActiveWithNow($afterSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isAfterScheduleActive, "Hour $hour: isAfterScheduleActive");
    }
  }

  public function testActiveStartEndMultiDay()
  {
    $now = Carbon::now();
    $zeroDay = Carbon::createFromTimestampMsUTC(0);

    for ($hour = 0; $hour < 24; $hour++) {

      $mockedNow = Carbon::create(
        $now->year,
        $now->month,
        $now->day,
        $hour,
        0,
        0,
      );

      $zeroDayWithMockedTime = $zeroDay->clone();
      $zeroDayWithMockedTime->setHours($hour);

      $beforeSchedule = new Schedule(
        $mockedNow->clone()->addDay()->timestamp * 1000,
        $mockedNow->clone()->addDays(2)->timestamp * 1000,
        'UTC',
        ScheduleTimeType::START_END,
        $zeroDayWithMockedTime->clone()->addHours(1)->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHours(3)->timestamp * 1000,
      );

      $isBeforeScheduleActive = Scheduler::isScheduleActiveWithNow($beforeSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isBeforeScheduleActive);

      $duringSchedule = new Schedule(
        $mockedNow->clone()->subDay()->timestamp * 1000,
        $mockedNow->clone()->addDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::START_END,
        $zeroDayWithMockedTime->clone()->subHour()->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHour()->timestamp * 1000,
      );

      $isDuringScheduleActive = Scheduler::isScheduleActiveWithNow($duringSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertTrue($isDuringScheduleActive);


      $AfterSchedule = new Schedule(
        $mockedNow->clone()->subDays(2)->timestamp * 1000,
        $mockedNow->clone()->subDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::START_END,
        $zeroDayWithMockedTime->clone()->subHours(3)->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->subHours(1)->timestamp * 1000,
      );

      $isAfterScheduleActive = Scheduler::isScheduleActiveWithNow($AfterSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isAfterScheduleActive);
    }
  }

  public function testActiveDailySingleDay()
  {
    $now = Carbon::now("UTC");
    $zeroDay = Carbon::createFromTimestampMsUTC(0);

    for ($hour = 0; $hour < 24; $hour++) {

      $mockedNow = Carbon::create(
        $now->year,
        $now->month,
        $now->day,
        $hour,
        0,
        0,
      );

      $zeroDayWithMockedTime = $zeroDay->clone();
      $zeroDayWithMockedTime->setHours($hour);

      $beforeWholeSchedule = new Schedule(
        $mockedNow->clone()->addDay()->timestamp * 1000,
        $mockedNow->clone()->addDays(3)->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->addHour()->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHours(3)->timestamp * 1000,
      );

      $isBeforeWholeScheduleActive = Scheduler::isScheduleActiveWithNow($beforeWholeSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isBeforeWholeScheduleActive, "Hour $hour: isBeforeWholeScheduleActive");

      $beforeDaySchedule = new Schedule(
        $mockedNow->clone()->subDay()->timestamp * 1000,
        $mockedNow->clone()->addDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->addHour()->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHours(3)->timestamp * 1000,
      );

      $isBeforeDayScheduleActive = Scheduler::isScheduleActiveWithNow($beforeDaySchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isBeforeDayScheduleActive, "Hour $hour: isBeforeDayScheduleActive");

      $duringDaySchedule = new Schedule(
        $mockedNow->clone()->subDay()->timestamp * 1000,
        $mockedNow->clone()->addDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->subHour()->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHour()->timestamp * 1000,
      );

      $isDuringDayScheduleActive = Scheduler::isScheduleActiveWithNow($duringDaySchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertTrue($isDuringDayScheduleActive, "Hour $hour: isDuringScheduleActive");


      $afterDaySchedule = new Schedule(
        $mockedNow->clone()->subDay()->timestamp * 1000,
        $mockedNow->clone()->addDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->subHours(3)->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->subHour()->timestamp * 1000,
      );

      $isAfterDayScheduleActive = Scheduler::isScheduleActiveWithNow($afterDaySchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isAfterDayScheduleActive, "Hour $hour: isAfterDayScheduleActive");


      $afterWholeSchedule = new Schedule(
        $mockedNow->clone()->subDays(3)->timestamp * 1000,
        $mockedNow->clone()->subDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->subHours(3)->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->subHour()->timestamp * 1000,
      );

      $isAfterWholeScheduleActive = Scheduler::isScheduleActiveWithNow($afterWholeSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isAfterWholeScheduleActive, "Hour $hour: isAfterWholeScheduleActive");
    }
  }

  public function testActiveDailyMultiDay()
  {
    $now = Carbon::now("UTC");
    $zeroDay = Carbon::createFromTimestampMsUTC(0);

    for ($hour = 0; $hour < 24; $hour++) {

      $mockedNow = Carbon::create(
        $now->year,
        $now->month,
        $now->day,
        $hour,
        0,
        0,
      );

      $zeroDayWithMockedTime = $zeroDay->clone();
      $zeroDayWithMockedTime->setHours($hour);

      $beforeWholeSchedule = new Schedule(
        $mockedNow->clone()->addDay()->timestamp * 1000,
        $mockedNow->clone()->addDays(3)->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->addHour()->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHours(3)->timestamp * 1000,
      );

      $isBeforeWholeScheduleActive = Scheduler::isScheduleActiveWithNow($beforeWholeSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isBeforeWholeScheduleActive, "Hour $hour: isBeforeWholeScheduleActive");

      $beforeDaySchedule = new Schedule(
        $mockedNow->clone()->subDay()->timestamp * 1000,
        $mockedNow->clone()->addDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->addHour()->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHours(3)->timestamp * 1000,
      );

      $isBeforeDayScheduleActive = Scheduler::isScheduleActiveWithNow($beforeDaySchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isBeforeDayScheduleActive, "Hour $hour: isBeforeDayScheduleActive");

      $duringDaySchedule = new Schedule(
        $mockedNow->clone()->subDay()->timestamp * 1000,
        $mockedNow->clone()->addDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->subHour()->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->addHour()->timestamp * 1000,
      );

      $isDuringDayScheduleActive = Scheduler::isScheduleActiveWithNow($duringDaySchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertTrue($isDuringDayScheduleActive, "Hour $hour: isDuringScheduleActive");


      $afterDaySchedule = new Schedule(
        $mockedNow->clone()->subDay()->timestamp * 1000,
        $mockedNow->clone()->addDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->subHours(3)->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->subHour()->timestamp * 1000,
      );

      $isAfterDayScheduleActive = Scheduler::isScheduleActiveWithNow($afterDaySchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isAfterDayScheduleActive, "Hour $hour: isAfterDayScheduleActive");

      $afterWholeSchedule = new Schedule(
        $mockedNow->clone()->subDays(3)->timestamp * 1000,
        $mockedNow->clone()->subDay()->timestamp * 1000,
        'UTC',
        ScheduleTimeType::DAILY,
        $zeroDayWithMockedTime->clone()->subHours(3)->timestamp * 1000,
        $zeroDayWithMockedTime->clone()->subHour()->timestamp * 1000,
      );

      $isAfterWholeScheduleActive = Scheduler::isScheduleActiveWithNow($afterWholeSchedule, ScheduleType::GLOBAL , $mockedNow);

      $this->assertFalse($isAfterWholeScheduleActive, "Hour $hour: isAfterWholeScheduleActive");
    }
  }
}

File: clients/csharp/Vexilla.Client.Tests/SchedulerTest.cs
using System;
using NUnit.Framework;
using System.Text.Json;

namespace Vexilla.Client.Tests
{
    [TestFixture]
    public class SchedulerTest
    {

        [Test]
        public void TestSchedulerActiveNone()
        {
            var schedule = new Schedule
            {
                Start = 0,
                End = 0,
                Timezone = "UTC",
                TimeType = ScheduleTimeType.None,
                StartTime = 0,
                EndTime = 0
            };

            Assert.True(
                Scheduler.IsScheduleActive(schedule, ScheduleType.Empty));
        }

        [Test]
        public void TestSchedulerActiveStartEndSingleDay()
        {
            var now = DateTimeOffset.UtcNow;
            var zeroDay = DateTimeOffset.FromUnixTimeMilliseconds(0);

            for(int hour = 0; hour < 24; hour++) {

                var mockedNow = new DateTimeOffset(
                    now.Year,
                    now.Month,
                    now.Day,
                    hour,
                    0,
                    0,
                    0,
                    TimeSpan.Zero
                );

                var zeroDayWithMockedTime = new DateTimeOffset(
                    zeroDay.Year,
                    zeroDay.Month,
                    zeroDay.Day,
                    hour,
                    0,
                    0,
                    0,
                    TimeSpan.Zero
                );

                var beforeSchedule = new Schedule
                {
                    Start = mockedNow.ToUnixTimeMilliseconds(),
                    End = mockedNow.ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.StartEnd,
                    StartTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(3).ToUnixTimeMilliseconds()
                };

                Assert.False(Scheduler.IsScheduleActiveWithNow(beforeSchedule,
                    ScheduleType.Global, mockedNow));

                var duringSchedule = new Schedule
                {
                    Start = mockedNow.ToUnixTimeMilliseconds(),
                    End = mockedNow.ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.StartEnd,
                    StartTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds()
                };

                Assert.True(Scheduler.IsScheduleActiveWithNow(duringSchedule,
                    ScheduleType.Global, mockedNow));

                var afterSchedule = new Schedule
                {
                    Start = mockedNow.ToUnixTimeMilliseconds(),
                    End = mockedNow.ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.StartEnd,
                    StartTime = zeroDayWithMockedTime.AddHours(-3).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds()
                };

                Assert.False(Scheduler.IsScheduleActiveWithNow(afterSchedule,
                    ScheduleType.Global, mockedNow));
            }
        }

        [Test]
        public void TestSchedulerActiveStartEndMultiDay()
        {
            var now = DateTimeOffset.Now;
            var zeroDay = DateTimeOffset.FromUnixTimeMilliseconds(0);

            for(int hour = 0; hour < 24; hour++) {

                var mockedNow = new DateTimeOffset(
                    now.Year,
                    now.Month,
                    now.Day,
                    hour,
                    0,
                    0,
                    0,
                    TimeSpan.Zero
                );

                var zeroDayWithMockedTime = new DateTimeOffset(
                    zeroDay.Year,
                    zeroDay.Month,
                    zeroDay.Day,
                    hour,
                    0,
                    0,
                    0,
                    TimeSpan.Zero
                );

                var beforeSchedule = new Schedule
                {
                    Start = mockedNow.AddDays(1).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(2).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.StartEnd,
                    StartTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(3).ToUnixTimeMilliseconds()
                };

                Assert.False(Scheduler.IsScheduleActiveWithNow(beforeSchedule,
                    ScheduleType.Global, mockedNow));

                var duringSchedule = new Schedule
                {
                    Start = mockedNow.AddDays(-1).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(1).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.StartEnd,
                    StartTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds()
                };

                Assert.True(Scheduler.IsScheduleActiveWithNow(duringSchedule,
                    ScheduleType.Global, mockedNow));

                var afterSchedule = new Schedule
                {
                    Start = mockedNow.AddDays(-2).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(-1).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.StartEnd,
                    StartTime = zeroDayWithMockedTime.AddHours(-3).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds()
                };

                Assert.False(Scheduler.IsScheduleActiveWithNow(afterSchedule,
                    ScheduleType.Global, mockedNow));
            }
        }

        [Test]
        public void TestSchedulingActiveDailySingleDay()
        {
            var now = DateTimeOffset.Now;
            var zeroDay = DateTimeOffset.FromUnixTimeMilliseconds(0);

            for(int hour = 0; hour < 24; hour++) {
                var mockedNow = new DateTimeOffset(
                    now.Year,
                    now.Month,
                    now.Day,
                    hour,
                    0,
                    0,
                    0,
                    TimeSpan.Zero
                );

                var zeroDayWithMockedTime = new DateTimeOffset(
                    zeroDay.Year,
                    zeroDay.Month,
                    zeroDay.Day,
                    hour,
                    0,
                    0,
                    0,
                    TimeSpan.Zero
                );

                var beforeWholeSchedule = new Schedule
                {
                    Start = mockedNow.AddDays(1).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(1).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(3).ToUnixTimeMilliseconds()
                };

                Assert.False(Scheduler.IsScheduleActiveWithNow(beforeWholeSchedule,
                    ScheduleType.Global, mockedNow));

                var beforeDaySchedule = new Schedule
                {
                    Start = mockedNow.ToUnixTimeMilliseconds(),
                    End = mockedNow.ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(3).ToUnixTimeMilliseconds()
                };

                Assert.False(Scheduler.IsScheduleActiveWithNow(beforeDaySchedule,
                    ScheduleType.Global, mockedNow));

                var duringSchedule = new Schedule
                {
                    Start = mockedNow.ToUnixTimeMilliseconds(),
                    End = mockedNow.ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds()
                };

                Assert.True(Scheduler.IsScheduleActiveWithNow(duringSchedule,
                    ScheduleType.Global, mockedNow));

                var afterDaySchedule = new Schedule
                {
                    Start = mockedNow.ToUnixTimeMilliseconds(),
                    End = mockedNow.ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(-3).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds()
                };
                Assert.False(Scheduler.IsScheduleActiveWithNow(afterDaySchedule,
                    ScheduleType.Global, mockedNow));

                var afterWholeSchedule = new Schedule
                {
                    Start = mockedNow.AddDays(-1).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(-1).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(-3).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds()
                };
                Assert.False(Scheduler.IsScheduleActiveWithNow(afterWholeSchedule,
                    ScheduleType.Global, mockedNow));
            }
        }


        [Test]
        public void TestSchedulingActiveDailyMultiDay()
        {
            var now = DateTimeOffset.Now;
            var zeroDay = DateTimeOffset.FromUnixTimeMilliseconds(0);

            for(int hour = 0; hour < 24; hour++) {
                var mockedNow = new DateTimeOffset(
                    now.Year,
                    now.Month,
                    now.Day,
                    hour,
                    0,
                    0,
                    0,
                    TimeSpan.Zero
                );

                var zeroDayWithMockedTime = new DateTimeOffset(
                    zeroDay.Year,
                    zeroDay.Month,
                    zeroDay.Day,
                    hour,
                    0,
                    0,
                    0,
                    TimeSpan.Zero
                );

                var beforeWholeSchedule = new Schedule
                {
                    Start = mockedNow.AddDays(1).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(3).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(3).ToUnixTimeMilliseconds()
                };

                Assert.False(Scheduler.IsScheduleActiveWithNow(beforeWholeSchedule,
                    ScheduleType.Global, mockedNow));

                var beforeDaySchedule = new Schedule
                {
                    Start = mockedNow.AddDays(-1).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(1).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(3).ToUnixTimeMilliseconds()
                };

                Assert.False(Scheduler.IsScheduleActiveWithNow(beforeDaySchedule,
                    ScheduleType.Global, mockedNow));


                var duringSchedule = new Schedule
                {
                    Start = mockedNow.AddDays(-1).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(1).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(1).ToUnixTimeMilliseconds()
                };

                Assert.True(Scheduler.IsScheduleActiveWithNow(duringSchedule,
                    ScheduleType.Global, mockedNow));

                var afterDaySchedule = new Schedule
                {
                    Start = mockedNow.AddDays(-1).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(1).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(-3).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds()
                };
                Assert.False(Scheduler.IsScheduleActiveWithNow(afterDaySchedule,
                    ScheduleType.Global, mockedNow));

                var afterWholeSchedule = new Schedule
                {
                    Start = mockedNow.AddDays(-3).ToUnixTimeMilliseconds(),
                    End = mockedNow.AddDays(-1).ToUnixTimeMilliseconds(),
                    Timezone = "UTC",
                    TimeType = ScheduleTimeType.Daily,
                    StartTime = zeroDayWithMockedTime.AddHours(-3).ToUnixTimeMilliseconds(),
                    EndTime = zeroDayWithMockedTime.AddHours(-1).ToUnixTimeMilliseconds()
                };
                Assert.False(Scheduler.IsScheduleActiveWithNow(afterWholeSchedule,
                    ScheduleType.Global, mockedNow));
            }
        }
    }
}